suprb/optimizer/rule/ns/ns.py

```diff
@@ -138,171 +120,67 @@ def _optimize(self, X: np.ndarray, y: np.ndarray, n_rules: int) -> list[Rule]:
-        """ Calculation of the Novelty Score based on the hamming distance.
-
-            Takes every rule and calculates the hamming distance to each rule contained in the archive and averages the
-            distances of the k-nearest neighbors to get the novelty score.
-
-            Return: A list of tuples containing each rule with its corresponding novelty score and also the amount of
-            matched data points, which is used as the tiebreaker in case of equal novelty score.
-        """
-        rules_with_novelty_score = []
-
-        # filter rules for minimal criteria
-        if self.ns_type == 'MCNS' and self.pool_:
-            rules = self._filter_for_minimal_criteria(rules)
-
-        if self.novelty_fitness_combination == 'pmcns':
-            rules = self._filter_for_progressive_minimal_criteria(rules)
-
-        valid_archive = list(filter(lambda r: r.is_fitted_ and r.experience_ > 0, archive))
-
-        # main loop with option for local competition
-        for rule in rules:
-            rules_w_distances = sorted([(B, hamming(rule.match_, B.match_)) for B in archive], key=lambda a: a[1])
-
-            distances = [a[1] for a in rules_w_distances]
-            novelty_score = sum(distances[:k]) / len(distances[:k])
-
-            # no need to enter if there is no valid_archive
-            if self.ns_type == 'NSLC' and valid_archive:
-                novelty_score = novelty_score + self._get_local_score(rule, rules_w_distances)
+                ns_children = self.novelty_calculation(rules=children)
+                ns_parents = self.novelty_calculation(rules=parents)
 
-            # matched_data_count is the secondary key which decides what rule is better if novelty is the same
-            matched_data_count = np.count_nonzero(rule.match_)
+                self.novelty_calculation.archive.extend_archive(ns_children, self.roh)
 
-            # linear combination of fitness and novelty
-            # any changes to fitness scaling would need to be also applied here
-            scaled_fitness = rule.fitness_ / 100
-            if self.novelty_fitness_combination == '50/50':
-                novelty_score = 0.5 * novelty_score + 0.5 * scaled_fitness
-            elif self.novelty_fitness_combination == '75/25':
-                novelty_score = 0.75 * novelty_score + 0.25 * scaled_fitness
+                best_children = self._get_n_best_rules(ns_children, self.mu)
+                best_parents = self._get_n_best_rules(ns_parents, self.n_elitists)
 
-            rules_with_novelty_score.append((rule, novelty_score, matched_data_count))
+                population = best_children + best_parents
 
-        if self.novelty_fitness_combination == 'pareto' and self.pool_:
-            rules_with_novelty_score = self._get_pareto_front(rules_with_novelty_score)
-
-        return rules_with_novelty_score
+        return [ns_rule for ns_rule in self.novelty_calculation.archive.get_archive() + best_children][:n_rules]
```

Comment:
we probably want to remove the novelty_score_ attribute from these rules as it is a meaningless value after we discard the archive

Reply:
`novelty_score_` is never added to the rule itself (although it might seem this way) as changing a loop variable only works if is accessed by index (see example below)

```
class A():
    def __init__(self):
        self.a = [1,2,3,4,5,6]
        
temp = A()

print("init", temp.a)

temp.b = 2

print("adding variable", temp.b)

for i in temp.a:
    i += 1
    
print("not working modification", temp.a)

for i in range(len(temp.a)):
    temp.a[i] += 1
    
print("working modification",temp.a)
```